build: context page (#16094)
docker/docs: build/building/context.md

---
title: Build context
description: Learn how to use the build context to access files from your Dockerfile
keywords: build, buildx, buildkit, context, git, tarball, stdin
---

The [`docker build`](../../engine/reference/commandline/build.md) or
[`docker buildx build`](../../engine/reference/commandline/buildx_build.md)
commands build Docker images from a [Dockerfile](../../engine/reference/builder.md)
and a "context".

A build's context is the set of files located in the specified `PATH` or `URL`.
The build process can refer to any of the files in the context. For example,
your build can use a [`COPY` instruction](../../engine/reference/builder.md#copy)
to reference a file in the context or a [`RUN --mount=type=bind` instruction](../../engine/reference/builder.md#run---mounttypebind)
for better performance with [BuildKit](../buildkit/index.md). The build context
is processed recursively. So, a `PATH` includes any subdirectories and the
`URL` includes the repository and its submodules.

## `PATH` context

> When building with [BuildKit](../buildkit/index.md), which sets up an
> "interactive" session, only requested files that are needed by the build from
> the client are sent over the entire build context. On the other hand, the
> legacy builder will send the entire context (recursively) to the daemon.

Reviewer: ```suggestion
When building with [BuildKit](../buildkit/index.md), the client only sends
the files that the daemon requires to complete the build. This is in contrast
with the legacy builder, where the client always sends the entire context,
recursively, to the daemon.
```

To increase the build's performance, exclude files and directories by adding
a [`.dockerignore` file](../../engine/reference/builder.md#dockerignore-file) to

> the context directory.

Reviewer: The previous paragraph makes it sound like we don't need to use a `.dockerignore` file if we are using BuildKit, because only files used in the build are sent over to the daemon anyway. What role does the `.dockerignore` file play in this context?

Author: > What role does the `.dockerignore` file play in this context?

With this instruction:

```dockerfile
COPY . .
```

All files will be sent anyway (BuildKit or legacy). The `.dockerignore` file can be useful to exclude some of them without having multi COPY instructions in your Dockerfile for the files/folders you need.

Also with BuildKit if you don't have `COPY` commands in the stages you are building there is no build context sent at all. If you have `COPY foo .` only file in build context for that build would be `foo` whereas legacy builder still send the whole context.

Author: I've added more explanation on this page, let me know if it seems clear enough.

#16 sha256:23ca2f94460dcbaf5b3c3edbaaa933281a4e0ea3d92fe295193e4df44dc68f85
#16 transferring context: 13.16MB 2.2s done
...
```

> **Warning**
>
> Avoid using your root directory, `/`, as the `PATH` for your build context,
> as it causes the build to transfer the entire contents of your hard drive to
> the daemon.
{:.warning}

## `URL` context

The `URL` parameter can refer to three kinds of resources:
* [Git repositories](#git-repositories)
* Pre-packaged [tarball contexts](#tarball-contexts)
* Plain [text files](#text-files)

### Git repositories

When the `URL` parameter points to the location of a Git repository, the
repository acts as the build context. The system recursively fetches the
repository and its submodules. The commit history is not preserved. A
repository is first pulled into a temporary directory on your host. After that
succeeds, the directory is sent to the daemon as the context. Local copy gives
you the ability to access private repositories using local user credentials,
VPN's, and so forth.

> **Note**
>
> If the `URL` parameter contains a fragment the system will recursively clone
> the repository and its submodules using a `git clone --recursive` command.

> Git URLs accept context configuration in their fragment section, separated by a

Reviewer: actually my suggestion here is not good. I think it should read something like, "Git URLs accept a context configuration parameter in the form of a URL fragment."

for better performance with [BuildKit](../buildkit/index.md). The build context
is processed recursively. So, a `PATH` includes any subdirectories and the
`URL` includes the repository and its submodules.

## `PATH` context

This example shows a build command that uses the current directory (`.`) as a
build context:

```console
$ docker build .
...
#16 [internal] load build context
#16 sha256:23ca2f94460dcbaf5b3c3edbaaa933281a4e0ea3d92fe295193e4df44dc68f85
#16 transferring context: 13.16MB 2.2s done
...
```

With the following Dockerfile:

```dockerfile
FROM busybox
WORKDIR /src
COPY foo .
```

And this directory structure:

```
.
├── Dockerfile
├── bar
├── foo
└── node_modules
```

The legacy builder sends the entire directory to the daemon, including `bar`
and `node_modules` directories, even though the `Dockerfile` does not use
them. When using [BuildKit](../buildkit/index.md), the client only sends the
files required by the `COPY` instructions, in this case `foo`.

In some cases you may want to send the entire context:

```dockerfile
FROM busybox
WORKDIR /src
COPY . .
```

> But exclude some files or directories. You can use a [`.dockerignore`](../../engine/reference/builder.md#dockerignore-file)
> file to exclude them from being sent like:

Reviewer 2: ```suggestion
You can use a [`.dockerignore`](../../engine/reference/builder.md#dockerignore-file) file to exclude some files or directories from being sent:
```

```gitignore
# .dockerignore
node_modules
bar
```

> **Warning**
>
> Avoid using your root directory, `/`, as the `PATH` for your build context,
> as it causes the build to transfer the entire contents of your hard drive to
> the daemon.
{:.warning}

## `URL` context

The `URL` parameter can refer to three kinds of resources:
* [Git repositories](#git-repositories)
* Pre-packaged [tarball contexts](#tarball-contexts)

> * Plain [text files](#text-files)

Reviewer 2: We also support pre-built binaries.

Maybe "plain text files" might be better as "regular files" in the Unix sense?

Reviewer: Hmm. What's an **irregular** file then?

Reviewer 2: Ugh nope, yup that comment doesn't make sense :smile: Why would we be using a binary as a context.

### Git repositories

When the `URL` parameter points to the location of a Git repository, the
repository acts as the build context. The builder recursively pulls the

> repository and its submodules. The pulls ignore the commit history. A

Reviewer 2: "The pulls ignore the commit history"

Can we add some clarification for this? Is this just removing the `.git` directory, or is it a shallow-clone, or something else?
